### DISCUSSION TITLE
Provide first-class support for open-source packages

### DISCUSSION THREAD

**Discussion by shalvah** [npm/feedback#220] [Registry]
Hi there,

I recently came across [another report](https://www.darkreading.com/application-security/malicious-code-injected-via-google-chrome-extension-highlights-app-risks/d/d-id/1340100) of an open-source Chrome extension that had a new (malicious) version published without the repo being updated. I say "another" because I recall this happening with an NPM package (can't find the link now).

So, I was wondering if NPM could provide better open-source guarantees. I'm thinking:
- When a user creates/edits a package, they can mark it as "open-source"
- If a package is marked as "open-source", the published code must match with what's in the repository; otherwise publishes will fail. 
- The open-source guarantee is displayed prominently on the website. Alternately, a security warning is added for non-open source packages with a repo link: "NPM cannot guarantee that the contents of this package match the contents of the repository."
- If a package is marked as"open-source", that value can't be changed unless in a major release.

I'm probably missing a few things, but I thought I'd float the idea first. My main assertion is that more visibility is needed into package contents, as people less familiar with the registry often assume that the repo *is* the package (especially if you're coming from PHP's Packagist, which operates more like this).

**Comment by ljharb** [comment]
One difficulty here is that every package with a build process won’t match the contents of the repo - since generated files shouldn’t be committed, and sometimes packages won’t include the original source.

**Reply by shalvah** [reply]
Yeah. Personally, I think generated files _should_ be committed for this exact reason, but not everyone agrees.😅

Still, I think having something here is better than nothing. Maybe start with packages that do have the dists checked in (the author will have to specify this somehow)?🤔

**Reply by ljharb** [reply]
I would be staunchly against anything that in any way encourages committing artifacts. That’s not a worthy tradeoff to me.

**Comment by MylesBorins** [comment]
I think there are two distinct asks that you have here

1. Make requirements regarding licenses of open source projects, specifically that they can only change the license in a Semver-Major bump
2. Create security requirements regarding contents of repo matching contents of the package

For the first request, we do not currently have any requirements in place regarding the semveryness of a change, and assuming that a project has the legal right to change their license it would feel like over stepping our place to enforce a Semver-Major change. We do have the most recent license shown on the landing page for a project. There is an [existing RFC](https://github.com/npm/rfcs/pull/182) to allow auditing licenses... this to me feels like a really good approach to the problem, as it could make sure that folks are notified when a license changes in a breaking way.

For the security requirements... as mentioned by @ljharb it would be pretty hard for us to enforce a 1:1 package:repo equivalence due to the amount of transpilation and build steps in the JavaScript ecosystem. One solution being explored would be if we could use something like an [SBOM](https://en.wikipedia.org/wiki/Software_bill_of_materials) as a way to keep track of the provenance of a package from commit to publish.
